In [9]:
cd ..

/Users/emekgozluklu/Desktop/IDP/Repositories/tutor


/Users/emekgozluklu/Desktop/IDP/Repositories/venv/lib/python3.11/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [10]:
prompt_few_show_with_explanations = """Determine whether the mistake in student's solution is a concentration error or a logical error. If the error type is LOGICAL, make sure that a mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly. If the error type is CONCENTRATION, make sure that the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps. Explain your reasoning in one-two sentences.

QUESTION: Tom has a red marble, a green marble, a blue marble, and three identical yellow marbles. How many different groups of two marbles can Tom choose?
CORRECT SOLUTION: There are two cases here: either Tom chooses two yellow marbles (1 result), or he chooses two marbles of different colors ($\\binom{4}{2}=6$ results). The total number of distinct pairs of marbles Tom can choose is $1+6=\\boxed{7}$.
STUDENT'S SOLUTION: Tom can choose $ c\\left(\\begin{array}{l} 6 \\ 2 \\end{array}\\right)=\\frac{6.5}{2}=\\boxed{15} $.
ERROR EXPLANATION: Yellow marbles are thought as unique by the student.
ERROR TYPE: The error caused from a missing knowledge of student who does not know how to deal with unique instances in probability questions, LOGICAL.

QUESTION: A Senate committee has 8 Republicans and 6 Democrats.  In how many ways can we form a subcommittee with 3 Republicans and 2 Democrats?'
CORRECT SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{3} = 56$ ways to choose the Republicans.  There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{2} = 15$ ways to choose the Democrats.  So there are $56 \\times 15 = \\boxed{840}$ ways to choose the subcommittee.
STUDENT'S SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{2} = 28$ ways to choose the Republicans. There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{3} = 20$ ways to choose the Democrats. So there are $28 \\times 20 = \\boxed{480}$ ways to choose the subcommittee.
ERROR EXPLANATION: The student mistakenly switched the number of spots for Republicans and Democrats, resulting in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.
ERROR TYPE: CONCENTRATION, student did not read the problem given carefully, ended up using the initial conditions given in the question incorrectly.

QUESTION: How many total days were there in the years 2001 through 2004?
CORRECT SOLUTION: Including 2001 and 2004, there are 4 years total.\nHowever, 2004 is a leap year.\nTherefore, the number of days is $4\\times 365 + 1 = \\boxed{1461}$.
STUDENT'S SOLUTION: A year is 365 days. There are 4 years from 2001 through 2004.$4\\times 365 = \\boxed{1460}$.
ERROR EXPLANATION: Student did not consider the leap year.
ERROR TYPE: LOGICAL, student did not consider the fact that one of every four year is leap year which has 1 extra day.


QUESTION: Find the greatest common divisor of 91 and 72.
CORRECT SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^3 \\cdot 3^2$, so gcd(91, 72) = $\\boxed{1}$.
STUDENT'S SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^2 \\cdot 13^1$, so gcd(91, 72) = $\\boxed{13}$.
ERROR EXPLANATION: Student makes a calculation error while performing facotrization.
ERROR TYPE: CONCENTRATION, student could not perform the operations correctly even though they know the basic rules of finding greatest common divisor of two numbers.


QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR EXPLANATION: explanation_input
ERROR TYPE: """

In [11]:
prompt_few_shot_no_explanation = """Determine whether the mistake in student's solution is a concentration error or a logical error. If the error type is LOGICAL, make sure that a mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly. If the error type is CONCENTRATION, make sure that the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps. Explain your reasoning in one-two sentences.

QUESTION: Tom has a red marble, a green marble, a blue marble, and three identical yellow marbles. How many different groups of two marbles can Tom choose?
CORRECT SOLUTION: There are two cases here: either Tom chooses two yellow marbles (1 result), or he chooses two marbles of different colors ($\\binom{4}{2}=6$ results). The total number of distinct pairs of marbles Tom can choose is $1+6=\\boxed{7}$.
STUDENT'S SOLUTION: Tom can choose $ c\\left(\\begin{array}{l} 6 \\ 2 \\end{array}\\right)=\\frac{6.5}{2}=\\boxed{15} $.
ERROR TYPE: The error caused from a missing knowledge of student who does not know how to deal with unique instances in probability questions, LOGICAL.


QUESTION: A Senate committee has 8 Republicans and 6 Democrats.  In how many ways can we form a subcommittee with 3 Republicans and 2 Democrats?'
CORRECT SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{3} = 56$ ways to choose the Republicans.  There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{2} = 15$ ways to choose the Democrats.  So there are $56 \\times 15 = \\boxed{840}$ ways to choose the subcommittee.
STUDENT'S SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{2} = 28$ ways to choose the Republicans. There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{3} = 20$ ways to choose the Democrats. So there are $28 \\times 20 = \\boxed{480}$ ways to choose the subcommittee.
ERROR TYPE: Student did not read the problem given carefully, ended up using the initial conditions given in the question incorrectly, CONCENTRATION.

QUESTION: How many total days were there in the years 2001 through 2004?
CORRECT SOLUTION: Including 2001 and 2004, there are 4 years total.\nHowever, 2004 is a leap year.\nTherefore, the number of days is $4\\times 365 + 1 = \\boxed{1461}$.
STUDENT'S SOLUTION: A year is 365 days. There are 4 years from 2001 through 2004.$4\\times 365 = \\boxed{1460}$.
ERROR TYPE: Student did not consider the fact that one of every four year is leap year which has 1 extra day, LOGICAL.


QUESTION: Find the greatest common divisor of 91 and 72.
CORRECT SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^3 \\cdot 3^2$, so gcd(91, 72) = $\\boxed{1}$.
STUDENT'S SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^2 \\cdot 13^1$, so gcd(91, 72) = $\\boxed{13}$.
ERROR TYPE: Student could not perform the operations correctly even though they know the basic rules of finding greatest common divisor of two numbers, CONCENTRATION.


QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR TYPE: """

In [12]:
two_step_prompt_1 = """Explain the mistake in the student's solution. State if the mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly. Or if the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps, or not using the preconditions given in the question properly. Explain in one-two sentences.

QUESTION: Tom has a red marble, a green marble, a blue marble, and three identical yellow marbles. How many different groups of two marbles can Tom choose?
CORRECT SOLUTION: There are two cases here: either Tom chooses two yellow marbles (1 result), or he chooses two marbles of different colors ($\\binom{4}{2}=6$ results). The total number of distinct pairs of marbles Tom can choose is $1+6=\\boxed{7}$.
STUDENT'S SOLUTION: Tom can choose $ c\\left(\\begin{array}{l} 6 \\ 2 \\end{array}\\right)=\\frac{6.5}{2}=\\boxed{15} $.
ERROR EXPLANATION: Yellow marbles are thought as unique by the student.


QUESTION: A Senate committee has 8 Republicans and 6 Democrats.  In how many ways can we form a subcommittee with 3 Republicans and 2 Democrats?'
CORRECT SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{3} = 56$ ways to choose the Republicans.  There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{2} = 15$ ways to choose the Democrats.  So there are $56 \\times 15 = \\boxed{840}$ ways to choose the subcommittee.
STUDENT'S SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{2} = 28$ ways to choose the Republicans. There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{3} = 20$ ways to choose the Democrats. So there are $28 \\times 20 = \\boxed{480}$ ways to choose the subcommittee.
ERROR EXPLANATION: The student mistakenly switched the number of spots for Republicans and Democrats, resulting in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.

QUESTION: How many total days were there in the years 2001 through 2004?
CORRECT SOLUTION: Including 2001 and 2004, there are 4 years total.\nHowever, 2004 is a leap year.\nTherefore, the number of days is $4\\times 365 + 1 = \\boxed{1461}$.
STUDENT'S SOLUTION: A year is 365 days. There are 4 years from 2001 through 2004.$4\\times 365 = \\boxed{1460}$.
ERROR EXPLANATION: Student did not consider the leap year.


QUESTION: Find the greatest common divisor of 91 and 72.
CORRECT SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^3 \\cdot 3^2$, so gcd(91, 72) = $\\boxed{1}$.
STUDENT'S SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^2 \\cdot 13^1$, so gcd(91, 72) = $\\boxed{13}$.
ERROR EXPLANATION: Student makes a calculation error while performing facotrization.


QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR EXPLANATION: """

two_step_prompt_2 = """Determine whether the mistake in student's solution is a CONCENTRATION error or a LOGICAL error. If the error type is LOGICAL, make sure that a mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly. If the error type is CONCENTRATION, make sure that the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps, or not using preconditions given in the question properly. Explain your reasoning in one-two sentences. State the error type in the output explicitly.

QUESTION: Tom has a red marble, a green marble, a blue marble, and three identical yellow marbles. How many different groups of two marbles can Tom choose?
CORRECT SOLUTION: There are two cases here: either Tom chooses two yellow marbles (1 result), or he chooses two marbles of different colors ($\\binom{4}{2}=6$ results). The total number of distinct pairs of marbles Tom can choose is $1+6=\\boxed{7}$.
STUDENT'S SOLUTION: Tom can choose $ c\\left(\\begin{array}{l} 6 \\ 2 \\end{array}\\right)=\\frac{6.5}{2}=\\boxed{15} $.
ERROR EXPLANATION: Yellow marbles are thought as unique by the student.
ERROR TYPE: The error caused from a missing knowledge of student who does not know how to deal with unique instances in probability questions, LOGICAL.

QUESTION: A Senate committee has 8 Republicans and 6 Democrats.  In how many ways can we form a subcommittee with 3 Republicans and 2 Democrats?'
CORRECT SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{3} = 56$ ways to choose the Republicans.  There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{2} = 15$ ways to choose the Democrats.  So there are $56 \\times 15 = \\boxed{840}$ ways to choose the subcommittee.
STUDENT'S SOLUTION: There are 8 Republicans and 3 spots for them, so there are $\\binom{8}{2} = 28$ ways to choose the Republicans. There are 6 Democrats and 2 spots for them, so there are $\\binom{6}{3} = 20$ ways to choose the Democrats. So there are $28 \\times 20 = \\boxed{480}$ ways to choose the subcommittee.
ERROR EXPLANATION: The student mistakenly switched the number of spots for Republicans and Democrats, resulting in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.
ERROR TYPE: Student did not read the problem given carefully, ended up using the initial conditions given in the question incorrectly, CONCENTRATION.

QUESTION: How many total days were there in the years 2001 through 2004?
CORRECT SOLUTION: Including 2001 and 2004, there are 4 years total.\nHowever, 2004 is a leap year.\nTherefore, the number of days is $4\\times 365 + 1 = \\boxed{1461}$.
STUDENT'S SOLUTION: A year is 365 days. There are 4 years from 2001 through 2004.$4\\times 365 = \\boxed{1460}$.
ERROR EXPLANATION: Student did not consider the leap year.
ERROR TYPE: Student did not consider the fact that one of every four year is leap year which has 1 extra day, LOGICAL.

QUESTION: Find the greatest common divisor of 91 and 72.
CORRECT SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^3 \\cdot 3^2$, so gcd(91, 72) = $\\boxed{1}$.
STUDENT'S SOLUTION: $91 = 7^1 \\cdot 13^1$ and $72 = 2^2 \\cdot 13^1$, so gcd(91, 72) = $\\boxed{13}$.
ERROR EXPLANATION: Student makes a calculation error while performing facotrization.
ERROR TYPE: Student could not perform the operations correctly even though they know the basic rules of finding greatest common divisor of two numbers, CONCENTRATION.

QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR EXPLANATION: explanation_input
ERROR TYPE: """

In [72]:
prompt_zero_shot = """You are a chatbot that finds and classifies the error in the students solution. There are only two possible outcome for the questions: LOGICAL and CONCENTRATION. If the mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly the error type is LOGICAL. If the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps, or not using basic preconditions given in the question properly error type is CONCENTRATION. Do not repeat any given information. Explain the error and determine the error type. State the error type explicitly in the end.

QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR EXPLANATION: """

In [14]:
def run(prompt, model, data, explanation=False):
    correct = 0
    preds = []
    incorrect = []
    
    for i, rec in tqdm(enumerate(data)):
        p = prompt
        p = p.replace('question_input', rec['record'].question)
        p = p.replace('correct_input', rec['record'].formatted_solution)
        p = p.replace('student_input', rec['incorrect_solution'])
        
        if explanation:
            p = p.replace('explanation_input', rec['error_explanation'])
            
        out = model.query(p)[0]
        preds.append(out)
        
        if rec['error_type'] in out.upper():
            correct += 1
        else:
            incorrect.append(i)

    return preds, correct, incorrect


def run_2step(prompt1, prompt2, model, data, explanation=False):
    
    correct = 0
    preds = []
    incorrect = []
    explanations = []
    predictions = []
    
    i = 0
    for rec in tqdm(data):
        p = prompt1
        p = p.replace('question_input', rec['record'].question)
        p = p.replace('correct_input', rec['record'].formatted_solution)
        p = p.replace('student_input', rec['incorrect_solution'])
        
        out1 = model.query(p)[0]
        explanations.append(out1)
    
        p = prompt2
        p = p.replace('question_input', rec['record'].question)
        p = p.replace('correct_input', rec['record'].formatted_solution)
        p = p.replace('student_input', rec['incorrect_solution'])
        p = p.replace('explanation_input', out1)
    
        out2 = model.query(p)[0]
        predictions.append(out2)
    
        if rec['error_type'] in out2.upper():
            correct += 1
        else:
            incorrect.append(i)
    return preds, correct, incorrect, explanations
    

In [15]:
from IPython.display import display, Math, Latex, clear_output
from src.model.GPT import GPTModel, COSTS
from src.datasets.MATH import MATH, MATHRecord
import datetime
import json
import random
import openai as oai
from openai import OpenAI
from copy import copy
from tqdm import tqdm
import itertools
import pickle

random.seed(52)

# Prepare dataset
dataset = MATH('data/math.jsonl')
dataset.records = [rec for rec in dataset.records if len(rec.formatted_solution) < 750]  # to cut the token
random.shuffle(dataset.records)

# Prepare model
gpt3 = GPTModel(model="gpt-3.5-turbo-0125", temperature=0.7, max_tokens=500)
gpt4 = GPTModel(model="gpt-4o", temperature=1, max_tokens=500)

gpt4.model = "gpt-4o"
COSTS['gpt-4o'] = {'input': 5, 'output': 15}

file_train = 'output/final_checked_fine_tuning.pkl'
file_eval = 'output/evaluation_dataset/eval_v2_full.pkl'

data_train = pickle.load(open(file_train, 'rb'))
data_eval = pickle.load(open(file_eval, 'rb'))


def show_t(out):
    display(Latex(r"$\textbf{Question " +str(out['record'].id) + ": }$"))
    display(Latex(r"$\textbf{Question: }$"))
    display(Latex(out['record'].question))
    display(Latex(r"$\textbf{Solution: }$"))
    display(Latex(out['record'].formatted_solution))
    display(Latex(r"$\textbf{Error type: }$" + out['error_type']))
    display(Latex(r"-------"))

    display(Latex((r"$\textbf{ERROR EXPLANATION: }$"  + out['error_explanation'])))
    display(Latex((r"$\textbf{STUDENT'S SOLUTION: }$"  + out['incorrect_solution'])))
    display(Latex("================================================"))


### Train dataset scores

In [26]:
errors = [i['error_type'] for i in data_train]

print('CONCENTRATION', errors.count('CONCENTRATION'))
print('LOGICAL', errors.count('LOGICAL'))
print(errors.count('LOGICAL') / len(data_train))

CONCENTRATION 49
LOGICAL 55
0.5288461538461539


In [35]:
prompt_zero_shot

"Determine whether the mistake in student's solution is a CONCENTRATION error or a LOGICAL error. If the error type is LOGICAL, make sure that a mistake is made in reasoning or in the application of a rule, even though the basic operations are performed correctly. If the error type is CONCENTRATION, make sure that the incorrect result comes from a lapse in attention leads to an oversight or a careless mistake in the calculations or steps, or not using preconditions given in the question properly. Explain your reasoning in one-two sentences. State the error type in the output explicitly.\n\n\nQUESTION: question_input\nCORRECT SOLUTION: correct_input\nSTUDENT'S SOLUTION: student_input\nERROR TYPE: "

In [13]:
preds_zero_shot_gpt3_train, correct_zero_shot_gpt3_train, incorrect_zero_shot_gpt3_train = run(prompt_zero_shot, gpt3, data_train)
correct_zero_shot_gpt3_train / len(data_train)  # 0.6153846153846154

0.6153846153846154

In [14]:
preds_few_shot_gpt3_train, correct_few_shot_gpt3_train, incorrect_few_shot_gpt3_train = run(prompt_few_shot_no_explanation, gpt3, data_train)
correct_few_shot_gpt3_train / len(data_train)  # 0.5961538461538461

104it [01:36,  1.08it/s]


0.5961538461538461

In [21]:
preds_2step_gpt3_train, correct_2step_gpt3_train, incorrect_2step_gpt3_train, explanations_2step_gpt3_train = run_2step(two_step_prompt_1, two_step_prompt_2, gpt3, data_train)
correct_2step_gpt3_train / len(data_train)  # 0.6442307692307693

100%|█████████████████████████████████████████████████████████████████████████████████| 104/104 [03:10<00:00,  1.83s/it]


0.6442307692307693

In [22]:
preds_zero_shot_gpt4_train, correct_zero_shot_gpt4_train, incorrect_zero_shot_gpt4_train = run(prompt_zero_shot, gpt4, data_train)
correct_zero_shot_gpt4_train / len(data_train)  # 0.7115384615384616

104it [04:02,  2.33s/it]


0.7115384615384616

### Evaluation dataset scores

In [27]:
errors = [i['error_type'] for i in data_eval]

print('CONCENTRATION', errors.count('CONCENTRATION'))
print('LOGICAL', errors.count('LOGICAL'))
print(errors.count('LOGICAL') / len(data_eval))

CONCENTRATION 20
LOGICAL 28
0.5833333333333334


In [23]:
preds_zero_shot_gpt3_eval, correct_zero_shot_gpt3_eval, incorrect_zero_shot_gpt3_eval = run(prompt_zero_shot, gpt3, data_eval)
correct_zero_shot_gpt3_eval / len(data_eval)  # 0.5416666666666666

48it [01:08,  1.43s/it]


0.5416666666666666

In [28]:
preds_few_shot_gpt3_eval, correct_few_shot_gpt3_eval, incorrect_few_shot_gpt3_eval = run(prompt_few_shot_no_explanation, gpt3, data_eval)
correct_few_shot_gpt3_eval / len(data_eval)  # 0.5833333333333334

48it [00:47,  1.00it/s]


0.5833333333333334

In [29]:
preds_2step_gpt3_test, correct_2step_gpt3_test, incorrect_2step_gpt3_test, explanations_2step_gpt3_test = run_2step(two_step_prompt_1, two_step_prompt_2, gpt3, data_eval)
correct_2step_gpt3_test / len(data_eval)  # 0.5833333333333334

100%|███████████████████████████████████████████████████████████████████████████████████| 48/48 [01:31<00:00,  1.90s/it]


0.5833333333333334

In [30]:
preds_zero_shot_gpt4_eval, correct_zero_shot_gpt4_eval, incorrect_zero_shot_gpt4_eval = run(prompt_zero_shot, gpt4, data_eval)
correct_zero_shot_gpt4_eval / len(data_eval)  # 0.6458333333333334

48it [01:51,  2.33s/it]


0.6458333333333334

### Fine-tuned model scores

In [6]:
gpt_ft_old = GPTModel(model="gpt-3.5-turbo-0125", max_tokens=1000)
gpt_ft_old.model = 'ft:gpt-3.5-turbo-0125:tum-sot-hctl::9Quieny6'

NameError: name 'GPTModel' is not defined

In [32]:
preds_zero_shot_ft_eval, correct_zero_shot_ft_eval, incorrect_zero_shot_ft_eval = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft_eval / len(data_eval)  # 0.6666666666666666

48it [01:01,  1.28s/it]


0.6666666666666666

In [34]:
preds_few_shot_ft_eval, correct_few_shot_ft_eval, incorrect_few_shot_ft_eval = run(prompt_few_shot_no_explanation, gpt_ft, data_eval)
correct_few_shot_ft_eval / len(data_eval)  # 0.625

48it [00:33,  1.42it/s]


0.625

---

### New Fine Tuning Results

In [16]:
gpt_ft = GPTModel(model="gpt-3.5-turbo-0125", max_tokens=1000, temperature=0.7)
gpt_ft.model = 'ft:gpt-3.5-turbo-0125:tum-sot-hctl::9RgClAy4'

In [17]:
preds_zero_shot_ft2_eval, correct_zero_shot_ft2_eval, incorrect_zero_shot_ft2_eval = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft2_eval / len(data_eval)  # 0.6666666666666666

48it [00:38,  1.23it/s]


0.5625

In [18]:
preds_zero_shot_ft2_eval2, correct_zero_shot_ft2_eval2, incorrect_zero_shot_ft2_eval_2 = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft2_eval / len(data_eval)  # 0.6666666666666666

48it [00:33,  1.42it/s]


0.5625

In [19]:
preds_zero_shot_ft2_eval3, correct_zero_shot_ft2_eval3, incorrect_zero_shot_ft2_eval_3 = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft2_eval / len(data_eval)  # 0.6666666666666666

48it [00:41,  1.17it/s]


0.5625

In [20]:
preds_zero_shot_ft2_eval4, correct_zero_shot_ft2_eval4, incorrect_zero_shot_ft2_eval_4 = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft2_eval / len(data_eval)  # 0.6666666666666666

48it [00:34,  1.41it/s]


0.5625

In [ ]:
preds_zero_shot_ft2_eval5, correct_zero_shot_ft2_eval5, incorrect_zero_shot_ft2_eval_5 = run(prompt_zero_shot, gpt_ft, data_eval)
correct_zero_shot_ft2_eval / len(data_eval)  # 0.6666666666666666

In [27]:

from collections import Counter
all_incorrects = incorrect_zero_shot_ft2_eval_4 + incorrect_zero_shot_ft2_eval_3 + incorrect_zero_shot_ft2_eval_2 + incorrect_zero_shot_ft2_eval

count = Counter(all_incorrects)
for i, j in count.most_common(20):
    print(i, j, data_eval[i]['error_type'])




3 4 LOGICAL
5 4 LOGICAL
8 4 CONCENTRATION
9 4 CONCENTRATION
11 4 CONCENTRATION
12 4 CONCENTRATION
22 4 LOGICAL
24 4 LOGICAL
26 4 CONCENTRATION
27 4 LOGICAL
31 4 CONCENTRATION
33 4 CONCENTRATION
35 4 CONCENTRATION
36 4 CONCENTRATION
44 4 CONCENTRATION
47 4 LOGICAL
28 3 LOGICAL
45 3 CONCENTRATION
32 2 CONCENTRATION
2 1 LOGICAL


In [30]:
conc = [8, 9, 11, 12, 26, 31, 33, 35, 36, 44]

In [52]:
for d in data_eval:
    if d['record'].id == 8636:
        show_t(d)

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

In [33]:
# 8 Error Explanation update: 'overlooks the fact that the price given is the proice of three boxes and uses it as the price of one box.'
# 9 Error type update: logical
# 11 Error Explanation update: 'confuses and wrongly uses the precondition given in the question as...'
# 12 Error Explanation update: 'mistake in calculation: 26 / 6 = 6'
# 26 Error Explanation update: 'overlooks the precondition given in the question'
# 31 Error type update: logical
# 33 No updaate
# 35 Error Explanation update: delete misinterpreting
# 36 Explanationsiz bu bulunmaz
# 44 Explanationsiz bulunmaz


for i in conc:
    print(i)
    show_t(data_eval[i])
    input()
    clear_output()

44


<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

KeyboardInterrupt: Interrupted by user

In [39]:
prompt_only_question = """QUESTION: question_input
CORRECT SOLUTION: correct_input
STUDENT'S SOLUTION: student_input
ERROR EXPLANATION: """

preds_oq_ft2_eval, correct_oq_ft2_eval, incorrect_oq_ft2_eval = run(prompt_only_question, gpt_ft, data_eval)
correct_oq_ft2_eval / len(data_eval)  # 0.6666666666666666

48it [01:28,  1.83s/it]


0.0

In [47]:
p, c, i = run(prompt_zero_shot, gpt_ft_old, data_train[:50])
c / 50  # 0.6666666666666666

50it [00:24,  2.03it/s]


2.1

In [49]:
p, c, i = run(prompt_zero_shot, gpt_ft, data_train[:50])
c / 20  # 0.6666666666666666

50it [00:35,  1.41it/s]


1.85

- We have created a new train dataset - 104 questions, cross checked, hand crafted
- Defined error types and characteristics better
- Rectified evaluation dataset considering the new definitions

- We will make another fine tuning round, we have checked every sample in the train dataset and optimized prompts.

- Feedback generation not done yet.
TODO:
- Combine the results, send e-mail to professor to set time for meeting.
- 

In [95]:
preds_zero_shot_ft2_eval, correct_zero_shot_ft2_eval, incorrect_zero_shot_ft2_eval = run(prompt_zero_shot, gpt_ft, data_train[20:30])
preds_zero_shot_ft3_eval, correct_zero_shot_ft3_eval, incorrect_zero_shot_ft3_eval = run(prompt_zero_shot, gpt_ft, data_train[20:30])
preds_zero_shot_ft4_eval, correct_zero_shot_ft4_eval, incorrect_zero_shot_ft4_eval = run(prompt_zero_shot, gpt_ft, data_train[20:30])

10it [00:18,  1.90s/it]
10it [00:12,  1.24s/it]
10it [00:10,  1.06s/it]


In [108]:
preds_zero_shot_ft2_eval, correct_zero_shot_ft2_eval, incorrect_zero_shot_ft2_eval = run(prompt_zero_shot, gpt_ft, data_eval)
preds_zero_shot_ft3_eval, correct_zero_shot_ft3_eval, incorrect_zero_shot_ft3_eval = run(prompt_zero_shot, gpt_ft, data_eval)
preds_zero_shot_ft4_eval, correct_zero_shot_ft4_eval, incorrect_zero_shot_ft4_eval = run(prompt_zero_shot, gpt_ft, data_eval)




48it [00:57,  1.20s/it]
48it [01:02,  1.29s/it]
48it [01:02,  1.30s/it]


In [115]:
# LABELS
# G: Good example, clear error type
# E: Example with error, has to change
# M: Could be both error types
labels = []
for q, p1, p2, p3 in zip(data_eval, preds_zero_shot_ft2_eval, preds_zero_shot_ft3_eval, preds_zero_shot_ft4_eval):
    if q['error_type'] == 'LOGICAL':
        continue
    show_t(q)
    print(p1)
    print(p2)
    print(p3)

    labels.append(input('LABEL: '))
    clear_output()

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

ERROR EXPLANATION: The student may have miscalculated the ratio of the diameters of the griddles, leading to an incorrect ratio of the areas. This would result in the student not scaling the amount of flour needed correctly, leading to an incorrect final answer.
ERROR TYPE: LOGICAL
ERROR EXPLANATION: The student may have incorrectly calculated the ratio of the diameters of the griddles, leading to an incorrect ratio of the areas and ultimately the amount of flour needed for Paul's waffle.
ERROR TYPE: LOGICAL
ERROR EXPLANATION: One potential error a student could make is incorrectly calculating the ratio of the diameters of the two griddles. This could lead to an incorrect ratio of the areas and ultimately the wrong amount of flour needed for Paul's waffle.
ERROR TYPE: LOGICAL


KeyboardInterrupt: Interrupted by user

In [112]:
# 10459
# 9185

len(labels)

17

In [113]:
for i, q, label in zip(range(17), data_eval[:17], labels):
    print(i, q['error_type'], labels[i])

0 LOGICAL -
1 CONCENTRATION -
2 LOGICAL G
3 LOGICAL G
4 LOGICAL G
5 LOGICAL -
6 CONCENTRATION Sil
7 CONCENTRATION hata turu degistir
8 CONCENTRATION Since the price of a lunch is 4.5$... ekleyip al
9 CONCENTRATION Logicala cevir, kullan
10 LOGICAL G
11 CONCENTRATION G
12 CONCENTRATION G
13 CONCENTRATION Logical hataya cevir, explanationi updatele
14 LOGICAL G, explanationi kisalt
15 LOGICAL G
16 LOGICAL G


In [123]:
#L 10
#C 3
# 44'te soru doğru explanation düzeltilir
# 36 10luk tabanda doğru işlem yapmış olmasına rağmen sorunun 11lik tabanda sorduğunu fark etmemiş basic preconditionı görmemiş açıklamada belirt
# 35'te basic preconditionda sıkıntı var bunu açıklamada belirt
# 33 açıklamayı değiştir
# 31 çıkar bunu
# 26 farklı ünite basic precondition olduğunu belirt
# 32 solution eksik gibi silinebilir
#42 error explanation yanlışşş
# 43 açıklama değişicek

for i in [2, 3, 4, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 26, 33, 35, 36, 42, 43, 44]:
    print(i)
    q = data_eval[i]
    show_t(q)
    input()
    clear_output()

13


<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

KeyboardInterrupt: Interrupted by user

In [ ]:
ftv2[2]['error_explanation'] = 'The student incorrectly identified the number of valid combinations without sufficiently accounting for each restricted pairing in groups of three.'

ftv2[3]['error_explanation'] = 'The student forgot to consider that the hairy-nosed wombats are indistinguishable when calculating the number of ways to arrange them.'

ftv2[4]['error_explanation'] = 'The student did not understand that a sandwich consisting any of the ingredients would cause allergic reaction and thought all ingredients have to be in the sandwich.'

ftv2[7]['error_explanation'] = 'The student did not think of excluding two jokers when considering the total number of cards between them.'
ftv2[7]['error_type'] = 'LOGICAL'

ftv2[8]['error_explanation'] = 'The student multiplied the total cost of 3 lunches by 5, confusing that was the cost for each lunch.'
ftv2[8]['incorrect_solution'] = 'Price of a lunch is $\$4.5$.' + ftv2[8]['incorrect_solution']

ftv2[9]['error_explanation'] = 'The student did not consider $0$ as a multiple of $17$.'
ftv2[9]['error_type'] = 'LOGICAL'

ftv2[10]['error_explanation'] = 'The student used the average formula incorrectly, without considering how many students received each mark.'

ftv2[11]['error_explanation'] = 'The student did not realize the number of horses are different in the precondition given in the question and in the question itself.'

ftv2[12]['error_explanation'] = 'The student miscalculated the number of tiles needed for the width of the floor by dividing 24 by 6 incorrectly.'

ftv2[13]['error_explanation'] = 'Despite factorizing the numbers correctly and stating that the highest power should be selected for each prime, student mistakenly overlooked the factorization of 4 and used 2 instead.'

ftv2[14]['error_explanation'] = 'The student misunderstood the concept of mean and incorrectly set up the equation.'

ftv2[15]['error_explanation'] = 'The student mistakenly combined $\sqrt{6}$ and $\sqrt{24}$ without simplifying $\sqrt{24}$ first.'

ftv2[16]['error_explanation'] = 'The student made an error by not knowing the formula for terminating decimals, resulting in an incorrect final decimal representation.'


In [94]:
selected = []
for i, q, label in zip(range(20), data_train[:20], labels):
    print(q['record'].id, q['error_type'], label)
    # if label == 'G':
    #     selected.append((i, q))
    # if len(selected) == 10:
    #     break

1780 CONCENTRATION M
1751 CONCENTRATION G
1755 CONCENTRATION E
1757 CONCENTRATION G
1766 CONCENTRATION E, reduce the level of explanation
1769 LOGICAL G
5574 LOGICAL G
5598 CONCENTRATION G
5554 CONCENTRATION G
5561 LOGICAL G
5552 LOGICAL M
5555 LOGICAL G
5562 CONCENTRATION M
4684 LOGICAL G
4724 CONCENTRATION G
4725 CONCENTRATION G
4728 CONCENTRATION G
4680 CONCENTRATION G
4705 CONCENTRATION Edit
4712 CONCENTRATION G


In [74]:
preds_zero_shot_ft2_eval

['The error could be in the initial count of the arrangements if all the letters are unique. If the student mistakenly calculates this count as 4! instead of 5!, it would lead to an incorrect final answer.\nERROR TYPE: CONCENTRATION',
 'ERROR EXPLANATION: The student may have mistakenly thought that the probability of pulling the prize out of the box was being asked for, rather than the probability of not pulling the prize out.\nERROR TYPE: CONCENTRATION',
 'POTENTIAL ERROR: One potential error a student could make is not taking into account that the cards are dealt without replacement. This could lead to an incorrect probability calculation.\nERROR TYPE: LOGICAL',
 'ERROR EXPLANATION: One potential error a student could make is miscalculating the number of ways in which one or both dice show 1 after finding the number of ways in which neither die shows 1. This could result in an incorrect probability calculation.\nERROR TYPE: CONCENTRATION',
 'POTENTIAL ERROR: One potential error a st

In [79]:
for (i, q) in selected:
    show_t(q)
    print(preds_zero_shot_ft2_eval[i])
    print(preds_zero_shot_ft3_eval[i])
    print(preds_zero_shot_ft4_eval[i])
    input()
    clear_output()

In [54]:
#1751 -> explanation update
#1769 -> delete
#5574 -> explanation update: dogru factorization yanlis sonuc
#5598 -> explanation update: soruda ve cozumde dogru belirtip yanlis kullanmis
#5561 -> concentration
#5555: The student may have mistakenly used the Pythagorean theorem to find the length of the third side without considering that it might not be the hypotenuse.
#4724: The student may have made a mistake in the calculation of the base 10 integer by incorrectly evaluating the powers of 3 and the coefficients.


<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

In [124]:
prompt = """Take the potential error sentence and turn it into a form that states that it is already happened.

Example:
Potential Error: The potential error that could be performed is if the student calculate summation of possible ways of selecting Republicans and Democrats instead of multiplication. This could result in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.
Actual Error: The student calculated the summation of possible ways of selecting Republicans and Democrats instead of using multiplication, resulting in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.

Potential Error: The potential error that could be performed is if the student mistakenly switches the number of spots for Republicans and Democrats. This could result in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.
Actual Error: The student confused the number of spots for Republicans and Democrats, resulting in selecting the wrong values for the combinations and ultimately getting an incorrect final answer.

Potential Error: potential_error_input
Actual Error: """
gpt3.query(prompt)
from tqdm import tqdm
actual_errors = []

for i in [28, 47]:
    q = data_eval[i]
    out = gpt3.query(prompt.replace('potential_error_input', q['error_explanation']))
    actual_errors.append(out[0])
    print(i, q['record'].id, out[0])

28 12473 The student mistakenly calculated the inverse of the matrix $\begin{pmatrix} 1 & -2 \\ 1 & 4 \end{pmatrix}$, resulting in an incorrect final answer for the matrix $\mathbf{M}$.
47 9654 The student misapplied the AM-GM inequality by forgetting the equality condition, resulting in an incorrect determination of the minimum value for the expression.


In [84]:
for (i, q) in selected:
    print(q['record'].id, end=', ')

1751, 1757, 1769, 5574, 5598, 5554, 5561, 5555, 4684, 4724, 

In [104]:
from copy import deepcopy
sub_ds = deepcopy([i for i in data_train if i['record'].id in [1751, 1757, 1769, 5574, 5598, 5554, 5561, 5555, 4684, 4724]])

sub_ds[0]["error_explanation"] = 'Student calculated the probability of pulling the prize out of the box instead of the probability of not pulling the prize out.'

sub_ds[1]["error_explanation"] =  'The student calculated the subtraction incorrectly, resulting in an incorrect calculation of the ways in which one or both dice show 1.'

sub_ds[3]["error_explanation"] = "Even though the student correctly factorized 91 and 72,  gcd(91, 72) is miscalculated  due to a weak grasp on the concept of gcd between two relatively prime numbers."

sub_ds[4]["error_explanation"] = "Even though the correct ratio is stated both in the question and in the beginning of student's solution, a wrong ratio is used in student's following calculations. This leads to an incorrect final answer."

sub_ds[5]["error_explanation"] = "The student miscalculated the final subtraction to find the desired angle."

sub_ds[6]['error_explanation'] = 'The student calculated the result of division operation incorrectly.'
sub_ds[6]['error_type'] = 'CONCENTRATION'

sub_ds[7]['error_explanation'] = 'The student mistakenly used the Pythagorean theorem to find the length of the third side without considering that it might not be the hypotenuse.'

sub_ds[8]['error_explanation'] = 'The student used the base transformation formula incorrectly starting from the power of 1 instead of 0, resulting in a wrong answer.'

sub_ds[9]['error_explanation'] = 'The student mistakenly calculated the powers of 3 incorrectly, which led to an incorrect sum and final result for the conversion to base 10.'

del sub_ds[2]


In [106]:
for i in sub_ds:
    print(i['error_type'])

CONCENTRATION
CONCENTRATION
LOGICAL
CONCENTRATION
CONCENTRATION
CONCENTRATION
LOGICAL
LOGICAL
CONCENTRATION


In [96]:
for question in sub_ds:
    if (question["record"].id == 1751):
        question["error_explanation"] = "Student calculated the probability of pulling the prize out of the box instead of the probability of not pulling the prize out."
    elif (question["record"].id == 1757):
        question["error_explanation"] = "The student calculated the subtraction incorrectly, resulting in an incorrect calculation of the ways in which one or both dice show 1."
    elif (question["record"].id == 1769):
        show_t(question)

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

<IPython.core.display.Latex object>

In [ ]:
for question in final_checked_fine_tuning:
    if (question["record"].id == 1751):
        question["error_explanation"] = "Student calculated the probability of pulling the prize out of the box instead of the probability of not pulling the prize out."
    elif (question["record"].id == 1757):
        question["error_explanation"] = "The student calculated the subtraction incorrectly, resulting in an incorrect calculation of the ways in which one or both dice show 1."
    elif (question["record"].id == 1769):
        # DELETE
    elif(question["record"].id == 5574):
        question["error_explanation"] = "Even though the student correctly factorized 91 and 72,  gcd(91, 72) is miscalculated  due to a weak grasp on the concept of gcd between two relatively prime numbers."
    elif(question["record"].id == 5598):
        question["error_explanation"] = "Even though the correct ratio is stated both in the question and in the beginning of student's solution, a wrong ratio is used in student's following calculations. This leads to an incorrect final answer."
    elif(question["record"].id == 5554):
        question["error_explanation"] = "The student miscalculated the final subtraction to find the desired angle."

In [3]:
cd ..

/Users/emekgozluklu/Desktop/IDP/Repositories/tutor


/Users/emekgozluklu/Desktop/IDP/Repositories/venv/lib/python3.11/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [4]:
import pickle
file_eval = 'output/evaluation_dataset/eval_v2_full.pkl'
data_eval = pickle.load(open(file_eval, 'rb'))

data_eval[2]['error_explanation'] = 'The student incorrectly identified the number of valid combinations without sufficiently accounting for each restricted pairing in groups of three.'

data_eval[3]['error_explanation'] = 'The student forgot to consider that the hairy-nosed wombats are indistinguishable when calculating the number of ways to arrange them.'

data_eval[4]['error_explanation'] = 'The student did not understand that a sandwich consisting any of the ingredients would cause allergic reaction and thought all ingredients have to be in the sandwich.'

data_eval[7]['error_explanation'] = 'The student did not think of excluding two jokers when considering the total number of cards between them.'
data_eval[7]['error_type'] = 'LOGICAL'

data_eval[8]['error_explanation'] = 'The student multiplied the total cost of 3 lunches by 5, confusing that was the cost for each lunch.'
data_eval[8]['incorrect_solution'] = 'Price of a lunch is $\$4.5$.' + data_eval[8]['incorrect_solution']

data_eval[9]['error_explanation'] = 'The student did not consider $0$ as a multiple of $17$.'
data_eval[9]['error_type'] = 'LOGICAL'

data_eval[10]['error_explanation'] = 'The student used the average formula incorrectly, without considering how many students received each mark.'

data_eval[11]['error_explanation'] = 'The student did not realize the number of horses are different in the precondition given in the question and in the question itself.'

data_eval[12]['error_explanation'] = 'The student miscalculated the number of tiles needed for the width of the floor by dividing 24 by 6 incorrectly.'

data_eval[13]['error_explanation'] = 'Despite factorizing the numbers correctly and stating that the highest power should be selected for each prime, student mistakenly overlooked the factorization of 4 and used 2 instead.'

data_eval[14]['error_explanation'] = 'The student misunderstood the concept of mean and incorrectly set up the equation.'

data_eval[15]['error_explanation'] = 'The student mistakenly combined $\sqrt{6}$ and $\sqrt{24}$ without simplifying $\sqrt{24}$ first.'

data_eval[16]['error_explanation'] = 'The student made an error by not knowing the formula for terminating decimals, resulting in an incorrect final decimal representation.'

data_eval[44]["error_explanation"] = "The student incorrectly distribute the minus sign after the subtracting the expression from the absolute value. This led to a wrong result."

data_eval[43]["error_explanation"] = "Student incorrectly computed $x$ value from the equation $3.44x = 24.08$ as $x = 6$ instead of correctly solving $x = \\frac{24.08}{3.44} = 7$, resulting in an incorrect value for x and consequently for y as well."
data_eval[43]['incorrect_solution'] = data_eval[43]['incorrect_solution'].replace('(6,6.4)$.', '\\boxed{(6,6.4)}$.')

data_eval[36]["error_explanation"] = "The student overlooked a basic precondition given in the question that the number of zeroes was needed to be calculated in base 11, not base 10."
data_eval[35]["error_explanation"] = "The student focused on the sum of the tens digits instead of the units digits, which doesn't fit with the basic precondition given in the question."
data_eval[33]["error_explanation"] = "The student neglected to consider the basic precondition that the number of planets must be less than 100, as specified in the question."

data_eval[26]['incorrect_solution'] = "Say Paul and Danny's griddles have diameters $d_1$ and $d_2$, respectively. Paul's griddle has $\\frac{d_1}{d_2}=\\frac{24\\text{ ft}}{6\\text{ ft}}=4$ times the diameter and therefore $\\frac{\\pi d_1^2/4}{\\pi d_2^2/4}=\\left(\\frac{r_1}{r_2}\\right)^2=4^2=16$ times the area. It requires $16$ times as much flour, or $16\\cdot0.5=\\boxed{8}$ cups of flour. Final answer: 8"
data_eval[26]["error_explanation"] = "The student did not realize that the diameters were given in different units which is a basic precondition of the question."

data_eval[47]["error_explanation"] = "The student misapplied the AM-GM inequality by not including the equality condition, resulting in an incorrect determination of the minimum value for the expression."
data_eval[28]["error_explanation"] = r"The student incorrectly calculated the inverse of the matrix $\begin{pmatrix} 1 & -2 \\ 1 & 4 \end{pmatrix}$ due to miscalculating the determinant of the given matrix, resulting in an incorrect final answer for the matrix $\mathbf{M}$."


In [6]:
data_eval[32]

{'record': MATH | Record: 10890 	 Question: There exist several positive integers $x$ such that $\frac{1}{x^2+x}$ is a terminating decimal. What is the second smallest such integer?,
 'error_type': 'CONCENTRATION',
 'error_explanation': 'A potential error that could be made while attempting to solve this question is mistakenly taking the first smallest integer instead of second.',
 'incorrect_solution': 'We can factor $\\frac{1}{x^2+x} = \\frac{1}{x(x+1)}$. either $x$ or $x+1$ is odd, so either $x$ or $x+1$ is a power of 5. We start by considering 5^0. If $x = 1$, we get $x+1 = 2$. we have the fraction $\\frac{1}{2}$, which terminates. The answer is is $x = \\boxed{1}$.',
 'annotator': 'Ufuk'}

In [7]:
selected = []
for i in [2, 3, 4, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 26, 33, 35, 36, 42, 43, 44, 28, 47]:
    selected.append(data_eval[i])

In [10]:
for s in selected:
    print(s['error_type'])

LOGICAL
LOGICAL
LOGICAL
LOGICAL
CONCENTRATION
LOGICAL
LOGICAL
CONCENTRATION
CONCENTRATION
CONCENTRATION
LOGICAL
LOGICAL
LOGICAL
CONCENTRATION
CONCENTRATION
CONCENTRATION
CONCENTRATION
LOGICAL
CONCENTRATION
CONCENTRATION
LOGICAL
LOGICAL


In [148]:
pickle.dump(selected, open('finetune_part2.pkl', 'wb'))

In [136]:
display(data_eval[26]['record'].solution)

"Say Paul and Danny's griddles have diameters $d_1$ and $d_2$, respectively. Paul's griddle has $\\frac{d_1}{d_2}=\\frac{24\\text{ ft}}{6\\text{ ft}}=4$ times the diameter and therefore $\\frac{\\pi d_1^2/4}{\\pi d_2^2/4}=\\left(\\frac{r_1}{r_2}\\right)^2=4^2=16$ times the area. It requires $16$ times as much flour, or $16\\cdot0.5=\\boxed{8}$ cups of flour. Final answer: 8"

In [141]:
data_eval[43]['incorrect_solution'].replace('(6,6.4)$.', '\\boxed{(6,6.4)}$.')

'First, we organize each equation by getting the variables on one side and the constants on the other. This makes our equations $2.2x -3.1y = -3.2$ and $0.4x + y = 8.8$. Solving the second equation for $y$ in terms of $x$ gives $y = 8.8-0.4x$. Substituting this into our other equation give \\begin{align*}&2.2x - 3.1(8.8-0.4x) = -3.2 \\\\ &2.2x - 27.28 + 1.24x = -3.2 \\\\ &3.44x = 24.08 \\\\ &x = 6.\\end{align*} So, $y = 8.8 - 0.4(6) = 6.4$, and our solution is $(x,y) = (6,6.4)$. Final answer: $(6,6.4)$'

In [22]:
selected[6]

{'record': MATH | Record: 11868 	 Question: A class of 30 students recently took a test.  If 20 students scored 80, 8 students scored 90, and 2 students scored 100, then what was the class average (the mean) on this test?,
 'error_type': 'LOGICAL',
 'error_explanation': 'The student used the average formula incorrectly, without considering how many students received each mark.',
 'incorrect_solution': 'The total amount orf marks is  $80+90+100=270$. Therefore the class average is $$ \\frac{270}{3} = \\boxed{90}.$$.',
 'annotator': 'Emek'}